prereise/gather/griddata/hifld/data_access/load.py

```diff
@@ -186,6 +189,102 @@ def get_hifld_electric_substations(path):
+
+        :param list segments: list of segments, each defined as a list of (lon, lat)
+            coordinate pairs.
+        :return: (*networkx.Graph*) -- a minimum spanning tree from the segments.
+        :raises ValueError: if the minimum spanning tree is not linear
+        """
+        # Build a Graph representing the existing connections between segment endpoints
+        g = nx.Graph()
+        # add a 'start' and 'end' node for each segment, and a zero-cost edge b/w them
+        for i in range(len(segments)):
+            g.add_edge((i, "start"), (i, "end"), weight=0)
+        # Add edge weights by the angular distance between different segment endpoints
+        for (seg1, which1), (seg2, which2) in itertools.combinations(g.nodes, 2):
+            if seg1 == seg2:
+                # no edge needed, since there's already a zero-cost edge there
+                continue
+            # Extract latitude and longitude from the appropriate end of each segment
+            lon1, lat1 = segments[seg1][0 if which1 == "start" else -1]
+            lon2, lat2 = segments[seg2][0 if which2 == "start" else -1]
+            edge_angle = angular_distance(
+                latlon_to_xyz(lat1, lon1), latlon_to_xyz(lat2, lon2)
+            )
+            g.add_edge((seg1, which1), (seg2, which2), weight=edge_angle)
+        # Identify edges to connect the segments
+        tree = nx.algorithms.tree.minimum_spanning_tree(g)
+        # Ensure that the minimum spanning tree is linear
+        if max(dict(nx.degree(tree)).values()) > 2:
+            raise ValueError("The minimum spanning tree isn't linear")
+        return tree
+
+    if len(segments) == 1:
+        # There's only one segment in the list, so no joining is necessary
+        return segments[0]
+
+    segments_df = pd.Series(segments).to_frame(name="segments")
+    segments_df["distance"] = segments_df["segments"].map(
+        lambda x: angular_distance(
+            latlon_to_xyz(x[0][1], x[0][0]),  # first point
+            latlon_to_xyz(x[-1][1], x[-1][0]),  # last point
+        )
+    )
+    filtered_segments = segments_df.loc[segments_df["distance"] > threshold]
+    if len(filtered_segments) == 1:
+        # There's only one segment left in the list, so no joining is necessary
+        return filtered_segments["segments"].iloc[0]
+
+    # Sort the segments by length, in case the minimum spanning tree isn't linear
+    sorted_segments = filtered_segments.sort_values("distance")
+    # Make a minimum spanning tree, dropping small segments as necessary for linearity
+    for i in range(len(sorted_segments)):
+        try:
+            linear_segments = sorted_segments.iloc[i:]["segments"].tolist()
+            tree = _generate_linear_spanning_tree(linear_segments)
+            # We found a good minimum spanning tree, so we can exit
+            break
+        except ValueError:
+            # The minimum spanning tree wasn't good, go to the next iteration
+            pass
```

Comment:
Another random thought to get more predictable results here: I'm wondering for invalid trees (nonlinear) whether it is always the case that there is main linear branch plus a few small branches along the way instead of multiple main branches. If so, probably we can remove nodes with `betweenness_centrality` being 1 which connect to a node with `betweenness_centrality` being greater or equal to 3.

Reply:
If there is a Y-shaped end of the main branch, i.e. the central node has `betweenness_centrality`== 3, we can drop the node with shorter edge (following the current convention). If it is in the middle, i.e. the central node has `betweenness_centrality` > 3, we simply drop all of them which shouldn't be a `tie` by definition. But if there is a Y-shaped structure in the middle, i.e. more than one main branches, this approach won't work since in the original process, we will keep removing edges until one main branch is completely removed.

Not sure this will cover all the corner cases though. If we are diving deep into a rabbit hole, let's stop here and keep whatever works as always:)

Reply:
I think you're right, in a three-segment non-linear connected tree, there is symmetry for the two 'longer' branches of the Y but there will be one that's shorter, just looking at graph topology.

I did a visual inspection on the latest dataset, and everything looks as reasonable as can be expected given an automated process and messy data. The logic for dropping the shortest one comes from this case (original on the left, joined on the right): 
![image](https://user-images.githubusercontent.com/7348392/163075366-eda39282-6603-4e9f-bf72-1a6af1496e9b.png)

It's hard to tell from visual inspection what the minimum spanning tree would look like, since all the ends are in roughly in the same location, but the short spur seems like the least important part of the line.

The worst output is actually a two-segment line, but one where once of the segments contains a doubled-back section, so the 'endpoint' is essentially in the middle:
![image](https://user-images.githubusercontent.com/7348392/163075468-b3c7d5d2-0938-40e7-b2b9-71a9063b5a8c.png)